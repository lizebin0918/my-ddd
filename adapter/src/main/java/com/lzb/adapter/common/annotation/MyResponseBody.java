package com.lzb.adapter.common.annotation;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * http通用Response，先不定义exclude字段，建议通过jackson注解处理<br/>
 * Created on : 2023-09-11 22:24
 * @author mac
 */
@Target({ElementType.TYPE, ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
@Documented
public @interface MyResponseBody {

}
