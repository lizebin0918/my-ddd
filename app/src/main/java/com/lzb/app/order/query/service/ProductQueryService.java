package com.lzb.app.order.query.service;

import java.util.List;
import java.util.Set;

import com.lzb.domain.order.aggregation.valobj.Sku;

public interface ProductQueryService {

    /**
     * 查询sku信息
     *
     * @param skuIds
     * @return
     */
    List<Sku> list(Set<Integer> skuIds);

}