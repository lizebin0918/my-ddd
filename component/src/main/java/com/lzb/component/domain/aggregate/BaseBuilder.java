package com.lzb.component.domain.aggregate;

/**
 * <br/>
 * Created on : 2023-09-12 13:32
 * @author lizebin
 */
public abstract class BaseBuilder<T extends BaseEntity<T>> {

    /*@NotNull
    @SuppressWarnings("unchecked")
    public Type getType() {
        return ((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[0];
    }

    public T getInstance() {
        return SpringHelper.getBean((Class<? extends T>) ((ParameterizedType)getType()).getRawType());
    }*/

    /**
     * 初始化对象
     * @return
     */
    public final T build() {
        T agg = doBuild();
        validate(agg);
        return agg;
    }

    /**
     * 只构造聚合根
     * @return
     */
    protected abstract T doBuild();

    /**
     * 子类重写，校验聚合根
     * @param entity
     */
    protected void validate(T entity) {

    }

}
