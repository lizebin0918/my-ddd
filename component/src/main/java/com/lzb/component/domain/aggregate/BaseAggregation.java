package com.lzb.component.domain.aggregate;


import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Objects;
import java.util.Queue;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.lzb.component.domain.event.DomainEvent;
import com.lzb.component.domain.exception.IllegalVersionException;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import lombok.extern.slf4j.Slf4j;

/**
 * 聚合根基类，包含共用属性和方法<br/>
 * <p>
 * Created on : 2022-02-24 22:49
 *
 * @author lizebin
 */
@Slf4j
@Getter
@SuperBuilder
public abstract class BaseAggregation<R extends BaseEntity<R>> extends BaseEntity<R> {

    /**
     * 领域事件
     */
    protected final transient Queue<DomainEvent> events = new LinkedList<>();

    /**
     * 订单的操作日志
     * 订单的操作日志(这里会有风险-类型不安全，暂时只能想到这个方法)
     */
    protected final transient List<OperationLog> logs = new ArrayList<>();

    /**
     * 快照组件
     */
    @JsonIgnore
    @Getter(AccessLevel.NONE)
    @Setter(AccessLevel.NONE)
    private final transient Snapshot<R> snapshot = new Snapshot<>();

    protected BaseAggregation(long id) {
        super(id);
    }

    /**
     * 切面设置快照
     */
    public void attachSnapshot() {
        snapshot.set((R) this);
    }

    /**
     * 获取快照
     *
     * @return
     */
    public R snapshot() {
        return snapshot.get();
    }

    /**
     * 清空快照
     */
    public void removeSnapshot() {
        snapshot.remove();
    }

    /**
     * 检查当前对象和快照版本，如果抛异常，表示当前线程获取两次聚合根，并且做了一次更新，版本号发生变化
     *
     * @throws IllegalVersionException
     */
    public void checkForVersion() throws IllegalVersionException {
        R currentSnapshot = snapshot.get();
        if (Objects.isNull(currentSnapshot)) {
            throw new IllegalVersionException("快照不存在，在查询聚合的方法上添加注解--@Snapshot");
        }
        if (!Objects.equals(this.version, currentSnapshot.getVersion())) {
            log.error("版本号有误：当前对象版本号={}，快照对象版本号={}", this.version, currentSnapshot.getVersion());
            throw new IllegalVersionException("快照版本号有误，请重试");
        }
    }

    /**
     * (领域事件)入队
     *
     * @param event
     */
    protected void addEvent(DomainEvent event) {
        events.add(event);
    }

    /**
     * 添加操作日志
     * @param operationLog 操作日志
     * @return boolean
     * @author mac
     * @date 2022/11/14
     */
    protected boolean addOperationLog(OperationLog operationLog) {
        return logs.add(operationLog);
    }

}
