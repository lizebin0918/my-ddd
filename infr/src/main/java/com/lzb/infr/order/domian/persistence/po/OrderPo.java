package com.lzb.infr.order.domian.persistence.po;

import java.math.BigDecimal;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import com.baomidou.mybatisplus.annotation.Version;
import com.lzb.domain.order.enums.OrderStatus;
import com.lzb.infr.common.BasePo;
import lombok.Data;
import lombok.EqualsAndHashCode;
import org.javers.core.metamodel.annotation.Id;

/**
 * <p>
 * 
 * </p>
 *
 * @author lizebin
 * @since 2023-08-29
 */
@Data
@EqualsAndHashCode(callSuper = false)
@TableName("\"order\"")
public class OrderPo extends BasePo {

    @Id
    @TableId(type = IdType.INPUT)
    private Long orderId;

    private OrderStatus orderStatus;

    private String currency;

    /**
     * 汇率
     */
    private BigDecimal exchangeRate;

    private BigDecimal totalShouldPay;

    private BigDecimal totalActualPay;

    @Version
    private int version;

    ///////////////////////////////////////////////////////////////////////////
    // 地址相关
    ///////////////////////////////////////////////////////////////////////////

    /**
     * 邮箱
     */
    private String email;

    /**
     * 电话
     */
    private String phoneNumber;

    private String firstName;

    private String lastName;

    /**
     * 地址1
     */
    private String addressLine1;

    /**
     * 地址2
     */
    private String addressLine2;

    /**
     * 国家
     */
    private String country;
}