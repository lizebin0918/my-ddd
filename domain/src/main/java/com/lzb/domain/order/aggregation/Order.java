package com.lzb.domain.order.aggregation;

import java.math.BigDecimal;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.lzb.component.domain.aggregate.BaseAggregate;
import com.lzb.component.exception.BizException;
import com.lzb.domain.order.dto.LockStockDto;
import com.lzb.domain.order.enums.OrderStatus;
import com.lzb.domain.order.event.OrderCanceledEvent;
import com.lzb.domain.order.event.OrderPlacedEvent;
import com.lzb.domain.order.valobj.FullName;
import com.lzb.domain.order.dto.SkuStockLock;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NonNull;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import one.util.streamex.StreamEx;

/**
 * 订单聚合根<br/>
 * 实体所有写的操作，必须通过聚合根调用，所以改成包可见(default)
 * 实体所有读操作，可以改成公共(public)
 * Created on : 2022-06-11 14:41
 *
 * @author lizebin
 */
@Slf4j
@Getter
// 方便测试构造
@Setter(AccessLevel.PACKAGE)
public class Order extends BaseAggregate<Order> {

    private OrderStatus orderStatus;

    private String currency;

    private BigDecimal exchangeRate;

    private BigDecimal totalShouldPay;

    private BigDecimal totalActualPay;

    private OrderAddress orderAddress;

    private OrderDetails orderDetails;

    protected Order(long id) {
        super(id);
    }

    /**
     * 构造器，@ConstructorProperties 用于反序列化
     * @param id
     * @param version
     * @param orderStatus
     * @param currency
     * @param exchangeRate
     * @param totalShouldPay
     * @param totalActualPay
     * @param orderAddress
     * @param orderDetails
     */
    @JsonCreator
    public Order(long id,
            int version,
            @NonNull OrderStatus orderStatus,
            @NonNull String currency,
            @NonNull BigDecimal exchangeRate,
            @NonNull BigDecimal totalShouldPay,
            @NonNull BigDecimal totalActualPay,
            @NonNull OrderAddress orderAddress,
            @NonNull OrderDetails orderDetails) {
        super(id);
        this.version = version;
        this.orderStatus = orderStatus;
        this.currency = currency;
        this.exchangeRate = exchangeRate;
        this.totalShouldPay = totalShouldPay;
        this.totalActualPay = totalActualPay;
        this.orderAddress = orderAddress;
        this.orderDetails = orderDetails;
    }

    public void updateAddress(OrderAddress newOrderAddress) {
        if (isShipped()) {
            throw new BizException("订单已发货，不能修改地址");
        }
        this.orderAddress = newOrderAddress;
    }

    public void updateSkuLockStock(List<SkuStockLock> skuStockLocks) {
        Map<Integer, SkuStockLock> skuId2Lock = skuStockLocks.stream().collect(Collectors.toMap(SkuStockLock::skuId, Function.identity()));
        Map<Integer, List<OrderDetail>> skuId2OrderDetails = StreamEx.of(orderDetails.toStream()).groupingBy(OrderDetail::getSkuId);
        skuId2OrderDetails.keySet().forEach(skuId -> {
            List<OrderDetail> skuOrderDetails = skuId2OrderDetails.getOrDefault(skuId, Collections.emptyList());
            for (int i = 1; i <= skuOrderDetails.size(); i++) {
                skuOrderDetails.get(i - 1).updateLocked(i <= skuId2Lock.get(skuId).lockedNum());
            }
        });
    }

    public void updateTotalActualPay(BigDecimal totalActualPay) {
        this.totalActualPay = totalActualPay;
    }

    /**
     * 取消订单
     */
    public void cancel() {
        orderStatus = orderStatus.toCancel();
        orderDetails.forEach(OrderDetail::cancel);
        addEvent(OrderCanceledEvent.create(id));
    }

    /**
     * 是否取消
     * @return
     */
    public boolean isCancel() {
        return orderStatus.isCancel();
    }

    /**
     * 生单
     */
    public void place() {
        addEvent(OrderPlacedEvent.create(this.id));
    }

    public boolean isShipped() {
        return false;
    }

    /**
     * 只更新姓名
     * @param fullName
     */
    public void updateFullName(FullName fullName) {
        this.orderAddress.updateFullName(fullName);
    }

    public boolean canCancel() {
        return false;
    }
}
