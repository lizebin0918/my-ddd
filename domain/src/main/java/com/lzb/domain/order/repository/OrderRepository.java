package com.lzb.domain.order.repository;

import java.util.List;

import com.lzb.component.domain.repository.CacheRepository;
import com.lzb.component.domain.repository.CommonRepository;
import com.lzb.domain.order.aggregation.Order;

/**
 * <br/>
 * Created on : 2023-08-30 23:07
 * @author mac
 */
public interface OrderRepository extends CommonRepository<Order>, CacheRepository<Order> {

    List<Order> list(long... orderIds);

}
