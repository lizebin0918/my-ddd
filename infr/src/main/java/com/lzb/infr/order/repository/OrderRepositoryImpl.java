package com.lzb.infr.order.repository;

import java.util.Collections;
import java.util.ConcurrentModificationException;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.function.LongSupplier;

import com.lzb.component.utils.MyDiff;
import com.lzb.component.utils.json.JsonUtils;
import com.lzb.domain.order.aggregation.Order;
import com.lzb.domain.order.repository.OrderRepository;
import com.lzb.infr.common.BaseRepository;
import com.lzb.infr.config.cache.CacheConstants;
import com.lzb.infr.order.converter.OrderConverter;
import com.lzb.infr.order.converter.OrderPoConverter;
import com.lzb.infr.order.persistence.po.OrderDetailPo;
import com.lzb.infr.order.persistence.po.OrderPo;
import com.lzb.infr.order.persistence.service.OrderDetailPoService;
import com.lzb.infr.order.persistence.service.OrderPoService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import one.util.streamex.LongStreamEx;
import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.tuple.ImmutablePair;

import org.springframework.cache.annotation.CacheConfig;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Repository;

/**
 * <br/>
 * Created on : 2023-08-30 23:08
 * @author mac
 */
@Slf4j
@Repository(OrderRepositoryImpl.BEAN_NAME)
@RequiredArgsConstructor(onConstructor_ = @Lazy)
@CacheConfig(cacheNames = {CacheConstants.ORDER})
public class OrderRepositoryImpl extends BaseRepository<Order> implements OrderRepository {

    public static final String BEAN_NAME = "OrderRepositoryImpl";

    private final OrderPoService orderPoService;

    private final OrderDetailPoService orderDetailPoService;

    @Override
    public LongSupplier doAdd(Order aggregate) {
        OrderPo orderPo = OrderPoConverter.toOrderPo(aggregate);
        orderPoService.save(orderPo);
        orderDetailPoService.saveBatch(OrderPoConverter.toOrderDetailPos(orderPo.getOrderId(), aggregate.getOrderDetails()));
        return orderPo::getOrderId;
    }

    @Override
    @CacheEvict(key = "#order.id")
    public Runnable doUpdate(Order order) {

        ImmutablePair<List<OrderDetailPo>, List<OrderDetailPo>> leftToAddAndRightToUpdate = diffOrderDetailPos(order);
        OrderPo orderPo = OrderPoConverter.toOrderPo(order);
        return () -> {
            updateByVersion(orderPo);
            saveAndUpdate(leftToAddAndRightToUpdate.getLeft(), leftToAddAndRightToUpdate.getRight());
        };
    }

    /**
     * diff出需要新增和更新的订单明细
     * @param order
     * @return
     */
    private static ImmutablePair<List<OrderDetailPo>, List<OrderDetailPo>> diffOrderDetailPos(Order order) {
        Order snapshot = order.snapshot();
        List<OrderDetailPo> currentOrderDetailPos = OrderPoConverter.toOrderDetailPos(order.getId(), order.getOrderDetails());
        List<OrderDetailPo> snapshotOrderDetailPos = OrderPoConverter.toOrderDetailPos(order.getId(), snapshot.getOrderDetails());
        return MyDiff.diff(OrderDetailPo.class, snapshotOrderDetailPos, currentOrderDetailPos);
    }

    private void saveAndUpdate(List<OrderDetailPo> addOrderDetailPos, List<OrderDetailPo> updateOrderDetailPos) {
        orderDetailPoService.saveBatch(addOrderDetailPos);
        orderDetailPoService.updateBatchById(updateOrderDetailPos);
    }

    private void updateByVersion(OrderPo orderPo) {
        boolean success = orderPoService.updateById(orderPo);
        // 只有订单才需要判断版本号
        if (!success) {
            log.info("根据版本号更新失败 {}", JsonUtils.toJSONString(orderPo));
            throw new ConcurrentModificationException("订单信息发生变化, 修改失败");
        }
    }

    @Override
    public Optional<Order> get(long id) {
        Optional<OrderPo> orderOpt = orderPoService.getOptById(id);
        if (orderOpt.isEmpty()) {
            return Optional.empty();
        }
        OrderPo orderPo = orderOpt.get();
        List<OrderDetailPo> orderDetailPos = orderDetailPoService.listByOrderId(id);
        return Optional.of(OrderConverter.toOrder(orderPo, orderDetailPos));
    }

    /**
     * 不缓存为null的值
     * @param id
     * @return
     */
    @Override
    @Cacheable(key = "#id", unless = "#result == null")
    public Optional<Order> getInCache(long id) {
        return get(id);
    }

    @Override
    public List<Order> list(long... orderIds) {
        if (ArrayUtils.isEmpty(orderIds)) {
            return Collections.emptyList();
        }
        Set<Long> orderIdSet = LongStreamEx.of(orderIds).boxed().toSet();
        List<OrderPo> orders = orderPoService.listByIds(orderIdSet);
        List<OrderDetailPo> orderDetails = orderDetailPoService.listByOrderIds(orderIdSet);
        return OrderConverter.toOrders(orders, orderDetails);
    }
}
