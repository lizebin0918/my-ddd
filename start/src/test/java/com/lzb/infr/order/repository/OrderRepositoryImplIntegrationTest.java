package com.lzb.infr.order.repository;

import java.math.BigDecimal;
import java.time.Clock;
import java.time.Instant;
import java.time.ZoneId;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import cn.hutool.core.util.IdUtil;
import com.baomidou.mybatisplus.core.metadata.TableInfoHelper;
import com.lzb.BaseIntegrationTest;
import com.lzb.component.helper.SpringHelper;
import com.lzb.domain.order.aggregate.Order;
import com.lzb.domain.order.repository.OrderRepository;
import com.lzb.infr.domain.order.persistence.po.OrderPo;
import com.lzb.infr.domain.order.repository.OrderRepositoryImpl;
import com.lzb.infr.event.persistence.DomainEventPo;
import com.lzb.infr.event.persistence.service.DomainEventPoService;
import jakarta.annotation.Resource;
import jakarta.validation.constraints.AssertTrue;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.MockedStatic;

import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.test.jdbc.JdbcTestUtils;

import static java.time.Instant.ofEpochMilli;

class OrderRepositoryImplIntegrationTest extends BaseIntegrationTest {

    @Resource(name = OrderRepositoryImpl.BEAN_NAME)
    private OrderRepository orderRepository;

    @Resource
    private DomainEventPoService domainEventPoService;

    @Resource
    private SpringHelper springHelper;

    @Resource
    private JdbcTemplate jdbcTemplate;

    @Test
    @DisplayName("测试SpringHelper")
    void should_get_bean_from_spring_helper() {
        assertThat(springHelper.getBean(OrderRepositoryImpl.BEAN_NAME)).isNotNull();
    }

    @Test
    @Sql("/sql/OrderRepositoryImplIntegrationTest/should_order_get.sql")
    @DisplayName("测试聚合根查询")
    void should_order_get() {

        long orderId = 1L;
        Order order = orderRepository.get(orderId).orElseThrow();
        assertThat(order.snapshot()).isNotNull();

        Map<String, Order> excepted = new HashMap<>();
        excepted.put("current", order);
        excepted.put("snapshot", order.snapshot());

        assertJSON(excepted);
    }

    @Test
    @Sql("/sql/OrderRepositoryImplIntegrationTest/should_update_order_info.sql")
    @DisplayName("测试聚合根查询")
    void should_update_order_info() {
        long orderId = 1L;
        Order order = orderRepository.get(orderId).orElseThrow();
        order.updateTotalActualPay(new BigDecimal("8888"));
        orderRepository.update(order);
        Order newOrder = orderRepository.get(orderId).orElseThrow();
        assertJSON(newOrder);
    }

    @Test
    @DisplayName("测试取消订单")
    @Sql("/sql/OrderRepositoryImplIntegrationTest/should_cancel_order.sql")
    void should_cancel_order() {
        // given
        long orderId = 1L;
        Order order = orderRepository.get(orderId).orElseThrow();
        Clock constantClock = Clock.fixed(ofEpochMilli(0), ZoneId.systemDefault());

        // DomainEvent设置了默认值，这里设置固定返回
        String msgId = "1";
        try (
            MockedStatic<Instant> instant = mockStatic(Instant.class);
            MockedStatic<IdUtil> idUtil = mockStatic(IdUtil.class);
        ) {
            instant.when(() -> Instant.now()).thenReturn(constantClock.instant());
            idUtil.when(() -> IdUtil.randomUUID()).thenReturn(msgId);
            order.cancel();
        }

        // when
        orderRepository.update(order);

        // then
        Order newOrder = orderRepository.get(orderId).orElseThrow();
        List<DomainEventPo> events = domainEventPoService.list();
        assertThat(events).hasSize(1);

        var assertMap = new HashMap<>();
        assertMap.put("event", events.get(0).getContent());
        assertMap.put("order", newOrder);
        assertJSON(assertMap);
    }

    @Test
    @DisplayName("测试缓存查询")
    @Sql("/sql/OrderRepositoryImplIntegrationTest/should_get_in_cache.sql")
    void should_get_in_cache() {

        long orderId = 1L;
        Order order1 = orderRepository.getInCache(orderId).orElseThrow();

        JdbcTestUtils.deleteFromTables(jdbcTemplate, TableInfoHelper.getTableInfo(OrderPo.class).getTableName());

        Order order2 = orderRepository.getInCache(orderId).orElseThrow();

        assertThat(order1.snapshot()).isNull();
        assertThat(order2.snapshot()).isNull();
        var assertMap = Map.of("order1", order1, "order2", order2);
        assertJSON(assertMap);

    }

    @Test
    @DisplayName("测试缓存清除")
    @Sql("/sql/OrderRepositoryImplIntegrationTest/should_clear_cache_when_get_in_cache.sql")
    void should_clear_cache_when_get_in_cache() {
        long orderId = 1L;
        Order order = orderRepository.getInCache(orderId).orElseThrow();
        assertThat(order).isNotNull();

        order = orderRepository.getOrThrow(orderId);
        order.cancel();
        orderRepository.update(order);

        Order cacheOrder = orderRepository.getInCache(orderId).orElseThrow();
        assertThat(cacheOrder.isCancel()).isTrue();
        assertThat(cacheOrder.getVersion()).isEqualTo(2);
        assertJSON(cacheOrder);
    }

    @Test
    @DisplayName("测试数据不存在的情况")
    void should_return_empty_when_order_not_found() {
        assertThat(orderRepository.getInCache(1L).isEmpty()).isTrue();

    }

}