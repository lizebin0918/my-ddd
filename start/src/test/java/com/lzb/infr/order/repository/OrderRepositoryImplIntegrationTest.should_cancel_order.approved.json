{
  "event" : "{\"key\": \"1\", \"tag\": \"ORDER_CANCELED\", \"bizId\": \"1\", \"msgId\": \"1\", \"orderId\": 1, \"timestamp\": 0}",
  "order" : {
    "currency" : "CNY",
    "exchangeRate" : 1,
    "id" : 1,
    "orderAddress" : {
      "country" : "CHINA",
      "email" : "123456@qq.com",
      "fullAddressLine" : {
        "addressLine1" : "line1",
        "addressLine2" : "line2"
      },
      "fullName" : {
        "firstName" : "first",
        "lastName" : "last"
      },
      "id" : 1,
      "phoneNumber" : "13800000000",
      "version" : 1
    },
    "orderDetails" : {
      "list" : [ {
        "id" : 1,
        "orderStatus" : "CANCELED",
        "price" : 1,
        "skuId" : 1,
        "version" : 1
      }, {
        "id" : 2,
        "orderStatus" : "CANCELED",
        "price" : 10,
        "skuId" : 2,
        "version" : 1
      } ]
    },
    "orderStatus" : "CANCELED",
    "totalActualPay" : 88,
    "totalShouldPay" : 88,
    "version" : 2
  }
}